File: guidebot_recorder/desktop/desktop.js
(() => {
  "use strict";

  // Role gating (Spec A), identical rationale to slide.js: context-level init
  // scripts run in every frame including the sandboxed site iframe, but the
  // desktop opener belongs only to the TOP document (the shell). Capture the
  // top-window check before any frame-bust neutralization can shadow it.
  const isTop = window === window.top;
  if (!isTop) {
    return;
  }

  const API_KEY = "__guidebot_desktop";
  const API_VERSION = 1;
  const ROOT_ID = "guidebot-desktop";
  const ROOT_SELECTOR = "[data-guidebot-desktop]";
  // Above the chrome shell (2147483644) so it covers the empty browser, but BELOW
  // the cursor (2147483647) and its click ripple ring (…646) so the arc and the
  // double-click stay visible on top of the desktop. Ties with the ripple disc
  // (…645), which is appended after this overlay, so DOM order keeps it on top.
  const Z_INDEX = "2147483645";

  const CFG = window.__guidebot_desktop_config || {};
  const BACKGROUND = CFG.background ?? "#1f3a63";
  const LABEL_COLOR = CFG.labelColor ?? "#f8fafc";
  const WINDOW_COLOR = CFG.windowColor ?? "#ffffff";
  const FONT_FAMILY =
    CFG.fontFamily ??
    "-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif";

  // The icon sits where a real desktop shortcut would: top-left quadrant, not
  // centred, so the cursor's arc to it reads as crossing the desktop.
  const ICON_LEFT_PCT = CFG.iconLeftPct ?? 12;
  const ICON_TOP_PCT = CFG.iconTopPct ?? 22;
  const ICON_SIZE = CFG.iconSize ?? 96;

  const previous = window[API_KEY];
  if (
    previous &&
    previous.__guidebotVersion === API_VERSION &&
    ["show", "hide", "ensure", "iconCenter", "openWindow", "token"].every(
      (name) => typeof previous[name] === "function"
    )
  ) {
    return;
  }

  let shownToken = 0;
  let lastDesktop = null;

  function setImportant(el, property, value) {
    el.style.setProperty(property, value, "important");
  }

  function mountRoot() {
    return document.documentElement || document.body;
  }

  function styleRoot(root) {
    root.id = ROOT_ID;
    root.setAttribute("data-guidebot-desktop", "");
    root.setAttribute("aria-hidden", "true");
    setImportant(root, "position", "fixed");
    setImportant(root, "inset", "0");
    setImportant(root, "margin", "0");
    setImportant(root, "border", "0");
    setImportant(root, "box-sizing", "border-box");
    setImportant(root, "background", lastDesktop?.color || BACKGROUND);
    setImportant(root, "font-family", FONT_FAMILY);
    setImportant(root, "z-index", Z_INDEX);
    // Hit-testable like the slide card (NOT pointer-events:none): a stray target
    // action during a desktop frame should fail Playwright's actionability check
    // rather than land on the page underneath.
    setImportant(root, "overflow", "hidden");
  }

  function buildIcon(desktop) {
    const wrap = document.createElement("div");
    wrap.setAttribute("data-guidebot-desktop-icon", "");
    setImportant(wrap, "position", "absolute");
    setImportant(wrap, "left", ICON_LEFT_PCT + "%");
    setImportant(wrap, "top", ICON_TOP_PCT + "%");
    setImportant(wrap, "transform", "translate(-50%, -50%)");
    setImportant(wrap, "display", "flex");
    setImportant(wrap, "flex-direction", "column");
    setImportant(wrap, "align-items", "center");
    setImportant(wrap, "gap", "10px");
    setImportant(wrap, "width", ICON_SIZE + 48 + "px");

    const glyph = document.createElement("div");
    setImportant(glyph, "width", ICON_SIZE + "px");
    setImportant(glyph, "height", ICON_SIZE + "px");
    setImportant(glyph, "display", "flex");
    setImportant(glyph, "align-items", "center");
    setImportant(glyph, "justify-content", "center");
    // Two icon sources, both trusted (never scenario free-text): a packaged
    // built-in arrives as an SVG string in `iconSvg`; the author's own file
    // arrives as a data URL in `iconImg`. Even so the SVG is adopted via
    // DOMParser + importNode rather than innerHTML — no HTML string is ever
    // assigned, so there is no injection surface to reason about.
    if (desktop && desktop.iconImg) {
      const img = document.createElement("img");
      img.src = desktop.iconImg;
      setImportant(img, "width", "100%");
      setImportant(img, "height", "100%");
      setImportant(img, "object-fit", "contain");
      glyph.appendChild(img);
    } else if (desktop && desktop.iconSvg) {
      const parsed = new DOMParser().parseFromString(desktop.iconSvg, "image/svg+xml");
      const svg = parsed.documentElement;
      if (svg && svg.nodeName.toLowerCase() === "svg" && !parsed.querySelector("parsererror")) {
        const adopted = document.importNode(svg, true);
        setImportant(adopted, "width", "100%");
        setImportant(adopted, "height", "100%");
        glyph.appendChild(adopted);
      }
    }
    wrap.appendChild(glyph);

    if (desktop && desktop.label) {
      const label = document.createElement("div");
      label.setAttribute("data-guidebot-desktop-label", "");
      setImportant(label, "color", LABEL_COLOR);
      setImportant(label, "font-size", "15px");
      setImportant(label, "font-weight", "500");
      setImportant(label, "text-align", "center");
      setImportant(label, "line-height", "1.25");
      setImportant(label, "text-shadow", "0 1px 3px rgba(0,0,0,0.45)");
      // TEXT, never markup — the label is author YAML.
      label.textContent = desktop.label;
      wrap.appendChild(label);
    }
    return wrap;
  }

  function unmount() {
    document.querySelectorAll(ROOT_SELECTOR).forEach((node) => node.remove());
  }

  function mount(desktop) {
    const parent = mountRoot();
    if (!parent) {
      return null;
    }
    unmount();
    const root = document.createElement("div");
    styleRoot(root);
    root.appendChild(buildIcon(desktop));
    parent.appendChild(root);
    return root;
  }

  function show(desktop) {
    lastDesktop = desktop ?? null;
    shownToken += 1;
    mount(lastDesktop);
    return shownToken;
  }

  function ensure(desktop) {
    if (desktop !== undefined) {
      lastDesktop = desktop ?? null;
    }
    const existing = document.querySelector(ROOT_SELECTOR);
    if (existing && existing.isConnected) {
      return existing;
    }
    return mount(lastDesktop);
  }

  function hide() {
    unmount();
  }

  // Viewport centre of the icon glyph, in CSS px — what the Python side moves the
  // real cursor to before the double-click. Read from layout, so it honours
  // whatever ICON_*_PCT resolved to at this viewport size.
  function iconCenter() {
    const wrap = document.querySelector("[data-guidebot-desktop-icon]");
    if (!wrap) {
      return null;
    }
    const rect = wrap.getBoundingClientRect();
    // The glyph is the first child; centre on it, not on the label below.
    const glyph = wrap.firstElementChild || wrap;
    const grect = glyph.getBoundingClientRect();
    return { x: grect.left + grect.width / 2, y: grect.top + grect.height / 2 };
  }

  // Grow a faux browser window from the icon to (almost) full frame. Purely
  // cosmetic: it bridges the double-click to the real chrome shell the Python
  // side reveals by hiding this overlay once the growth is done.
  function openWindow(ms) {
    const root = document.querySelector(ROOT_SELECTOR);
    const center = iconCenter();
    if (!root || !center) {
      return false;
    }
    const win = document.createElement("div");
    win.setAttribute("data-guidebot-desktop-window", "");
    const bar = document.createElement("div");
    setImportant(bar, "height", "34px");
    setImportant(bar, "background", "#e8ebf0");
    setImportant(bar, "border-radius", "10px 10px 0 0");
    setImportant(bar, "box-shadow", "inset 0 -1px 0 rgba(0,0,0,0.08)");
    win.appendChild(bar);
    setImportant(win, "position", "absolute");
    setImportant(win, "background", WINDOW_COLOR);
    setImportant(win, "border-radius", "10px");
    setImportant(win, "box-shadow", "0 24px 80px rgba(0,0,0,0.45)");
    setImportant(win, "overflow", "hidden");
    setImportant(win, "transform-origin", "top left");
    root.appendChild(win);

    const vw = window.innerWidth;
    const vh = window.innerHeight;
    // Final window: a small even margin around the frame.
    const margin = Math.round(Math.min(vw, vh) * 0.04);
    const fromW = 8;
    const fromH = 6;
    const start = {
      left: center.x - fromW / 2,
      top: center.y - fromH / 2,
      width: fromW,
      height: fromH,
    };
    const end = {
      left: margin,
      top: margin,
      width: vw - margin * 2,
      height: vh - margin * 2,
    };
    for (const key of ["left", "top", "width", "height"]) {
      setImportant(win, key, start[key] + "px");
    }
    if (typeof win.animate === "function") {
      win.animate(
        [
          { left: start.left + "px", top: start.top + "px", width: start.width + "px", height: start.height + "px", opacity: 0.6 },
          { left: end.left + "px", top: end.top + "px", width: end.width + "px", height: end.height + "px", opacity: 1 },
        ],
        { duration: Math.max(1, ms), easing: "cubic-bezier(0.16, 1, 0.3, 1)", fill: "forwards" }
      );
    }
    for (const key of ["left", "top", "width", "height"]) {
      setImportant(win, key, end[key] + "px");
    }
    return true;
  }

  function token() {
    return shownToken;
  }

  const api = {
    __guidebotVersion: API_VERSION,
    show,
    hide,
    ensure,
    iconCenter,
    openWindow,
    token,
  };

  Object.defineProperty(window, API_KEY, {
    configurable: true,
    enumerable: false,
    writable: true,
    value: api,
  });
})();

File: guidebot_recorder/desktop/desktop.py
"""Python controller for the full-frame desktop-opener overlay.

Mirrors :class:`guidebot_recorder.slide.slide.Slide`: a DOM overlay injected via
a Playwright init script and driven from Python through ``page.evaluate``. It
paints a coloured desktop with a single browser icon, exposes the icon's centre
so the caller can move the real cursor there, and animates a window growing out
of the icon — the visual half of the :class:`~guidebot_recorder.models.scenario.Desktop`
step. The other half (the cursor arc and double-click) is driven by the render
loop through the existing cursor overlay.

Re-exported as ``DesktopOverlay`` from :mod:`guidebot_recorder.desktop` to mirror
``SlideOverlay`` and avoid colliding with the scenario ``Desktop`` step model.
"""

from __future__ import annotations

import base64
import json
import mimetypes
from importlib.resources import files
from pathlib import Path
from typing import Any

from playwright.async_api import BrowserContext, Page

from guidebot_recorder.models.scenario import DESKTOP_ICON_ALIASES, Desktop

_API_IS_READY = """() => {
    const api = window.__guidebot_desktop;
    return !!api && ["show", "hide", "ensure", "iconCenter", "openWindow", "token"].every(
        (name) => typeof api[name] === "function"
    );
}"""

#: Image types a scenario may point ``desktop.icon`` at. SVG is included: it is
#: rendered as an <img> data URL, not adopted into the DOM, so it cannot script.
_ICON_SUFFIXES = {".svg", ".png", ".jpg", ".jpeg", ".gif", ".webp"}


def _builtin_icon_svg(name: str) -> str:
    """Return the packaged SVG source for a built-in icon *name* (already aliased)."""
    asset = DESKTOP_ICON_ALIASES[name]
    return (
        files("guidebot_recorder.desktop.icons")
        .joinpath(f"{asset}.svg")
        .read_text(encoding="utf-8")
    )


def _file_icon_data_url(path: Path) -> str:
    """Encode an author-supplied icon file as a ``data:`` URL for an <img> src."""
    suffix = path.suffix.lower()
    if suffix not in _ICON_SUFFIXES:
        raise ValueError(
            f"nieobsługiwany format ikony pulpitu {path.name!r}; "
            f"dozwolone: {', '.join(sorted(_ICON_SUFFIXES))}"
        )
    mime = mimetypes.types_map.get(suffix, "image/svg+xml" if suffix == ".svg" else None)
    if mime is None:  # pragma: no cover - defensive; every suffix above maps
        raise ValueError(f"nie rozpoznano typu MIME ikony {path.name!r}")
    encoded = base64.b64encode(path.read_bytes()).decode("ascii")
    return f"data:{mime};base64,{encoded}"


def resolve_icon(desktop: Desktop, *, base_dir: Path | None = None) -> dict[str, str]:
    """Turn a step's ``icon`` into the payload the JS overlay expects.

    A built-in name yields ``{"iconSvg": <packaged svg>}``; a path yields
    ``{"iconImg": <data url>}``. Relative paths resolve against *base_dir* (the
    scenario file's directory) so a scenario can ship its own icon beside it.

    Fail-loud on a bad path or unknown built-in: a desktop step naming an icon
    that cannot be drawn is an authoring error, not something to paper over with
    a blank square.
    """
    icon = desktop.icon
    if desktop.is_builtin_icon():
        return {"iconSvg": _builtin_icon_svg(icon)}
    path = Path(icon)
    if not path.is_absolute() and base_dir is not None:
        path = base_dir / path
    if not path.is_file():
        raise ValueError(
            f"ikona pulpitu {icon!r} nie jest ani wbudowaną nazwą "
            f"({', '.join(sorted(set(DESKTOP_ICON_ALIASES)))}) ani istniejącym plikiem"
        )
    return {"iconImg": _file_icon_data_url(path)}


class DesktopOverlay:
    """Install and control the desktop-opener overlay on a Playwright page.

    Named ``DesktopOverlay`` rather than ``Desktop`` (as ``Slide`` is) because
    this module also imports the scenario ``Desktop`` step model for
    :func:`resolve_icon`; the two must not collide.
    """

    def __init__(self, config: dict[str, Any] | None = None) -> None:
        self.config = config or {}
        body = files("guidebot_recorder.desktop").joinpath("desktop.js").read_text(encoding="utf-8")
        appearance = {
            "background": self.config.get("background", "#1f3a63"),
            "labelColor": "#f8fafc",
            "windowColor": "#ffffff",
        }
        prelude = f"window.__guidebot_desktop_config = {json.dumps(appearance)};\n"
        self._script = prelude + body

    async def install(self, page: Page) -> None:
        await page.add_init_script(script=self._script)
        await page.evaluate(self._script)

    async def install_context(self, context: BrowserContext) -> None:
        await context.add_init_script(script=self._script)

    async def _ensure_api(self, page: Page) -> None:
        if not await page.evaluate(_API_IS_READY):
            await page.evaluate(self._script)

    async def show(self, page: Page, desktop: dict[str, Any]) -> None:
        """Paint the desktop with ``desktop = {color, label, iconSvg|iconImg}``."""
        await self._ensure_api(page)
        await page.evaluate("d => window.__guidebot_desktop.show(d)", desktop)

    async def ensure(self, page: Page, desktop: dict[str, Any]) -> None:
        await self._ensure_api(page)
        await page.evaluate("d => window.__guidebot_desktop.ensure(d)", desktop)

    async def icon_center(self, page: Page) -> tuple[float, float] | None:
        """Viewport centre of the icon glyph, or ``None`` if no desktop is up."""
        await self._ensure_api(page)
        point = await page.evaluate("() => window.__guidebot_desktop.iconCenter()")
        if not isinstance(point, dict):
            return None
        return float(point["x"]), float(point["y"])

    async def open_window(self, page: Page, ms: float) -> bool:
        """Animate the window growing from the icon over *ms* milliseconds."""
        await self._ensure_api(page)
        return bool(await page.evaluate("m => window.__guidebot_desktop.openWindow(m)", float(ms)))

    async def hide(self, page: Page) -> None:
        await self._ensure_api(page)
        await page.evaluate("() => window.__guidebot_desktop.hide()")

    async def token(self, page: Page) -> Any:
        await self._ensure_api(page)
        return await page.evaluate("() => window.__guidebot_desktop.token()")
